eng/Tools.props

```diff
@@ -47,6 +47,15 @@
     <PackageReference Include="System.Security.Cryptography.Hashing.Algorithms.Analyzers" Version="1.1.0" />
   </ItemGroup>
 
+  <!-- Include the Roslyn Compilers -->
```

Comment:
Any chance we can restore *both*?  I would have expected the targets inside these packages to be smart enough to no-op when imported on the wrong MSBuild.  Perhaps that's wishful thinking.

Reply:
An alternative approach would be to restore with ExcludeAssets="Build" then manually construct the imports based on condition in Directory.Build.props|targets.

Reply:
No, that is what I was doing originally, but the packages are not smart enough to handle this.

The compiler expects the consumer to pick the appropriate version based on the compiler that is doing the build (which works with most solutions, since they rely on the regular per-project restore).

Reply:
> An alternative approach would be to restore with ExcludeAssets="Build"

That might work, but that also starts getting "fuzzy" when you need to determine where the package was restored to.

Reply:
> That might work, but that also starts getting "fuzzy" when you need to determine where the package was restored to.

But it might work better on the sense that you wouldn't have to run restore on Tools.proj with full MSBuild to be able to use latest compiler feature if you want to build an individual project with full MSBuild.

Right now, when calling `build.cmd` it uses `dotnet msbuild` by default, so it will get dotnet's compilers.

Reply:
This needs to get fixed, IMHO, Switching packagereferences based on msbuildruntimetype is broken if you do VS restores and dotnet build with implicit back and forth. I have been bitten by this. MSBuildRuntimeType. I've advocated for one package in the past that just works on both. cc @jaredpar

Reply:
> Right now, when calling `build.cmd` it uses `dotnet msbuild` by default, so it will get dotnet's compilers.

Which aren't what we want in this case. We are currently using (by default) the 2.1.400 SDK, which comes with the 2.9.0 compiler. We want the 3.0.0-preview compiler, regardless of what `dotnet` you call (whether our restored copy or a local one on your path).

> But it might work better on the sense that you wouldn't have to run restore on Tools.proj with full MSBuild to be able to use latest compiler feature if you want to build an individual project with full MSBuild

The problem is that we need to determine where `Microsoft.Net.Compilers` and `Microsoft.NetCore.Compilers` are restored to on-box. They could be restored to an explicit location or they could already exist on-box in the user's NuGet cache (which they can and sometimes do relocate from the default location).
It basically becomes a set of non-trivial checks to locate the appropriate packages (something which NuGet normally does for us).

In the end, it seems like more effort than its worth (at least to me) and that this time would be better spent getting CoreFX on a per-project restore system. That should actually be feasible with the various perf-fixes that have come about for large solution restore and the existence of solution filters in VS 2019; the caveat would be that you couldn't easily use full MSBuild to build until after the work was complete.

Reply:
> This needs to get fixed, IMHO. Switching packagereferences based on msbuildruntimetype is broken if you do VS restores and dotnet build with implicit restore back and forth. I have been bitten by this. I've advocated for one package in the past that just works on both. cc @jaredpar

I don't disagree and actually asked @agocke about this at lunch today. He said something to the effect of not wanting to bloat the package with the duplicate binaries. A meta-package might solve that issue, for people who want it.

Reply:
> The problem is that we need to determine where `Microsoft.Net.Compilers` and `Microsoft.NetCore.Compilers` are restored to on-box. They could be restored to an explicit location or they could already exist on-box in the user's NuGet cache (which they can and sometimes do relocate from the default location).
It basically becomes a set of non-trivial checks to locate the appropriate packages (something which NuGet normally does for us).

You could technically use NugetPackageRoot which is set in arcade. Therefore it will be available for corefx in Directory.Build.props.

https://github.com/dotnet/arcade/search?q=NuGetPackageRoot&unscoped_q=NuGetPackageRoot

Reply:
I'm going to go as far as saying having different packages for different msbuild runtimes is an anti-pattern. It's incorrect because it's impossible to have a good experience using both `dotnet` and `VS` side by side. Nothing else that ships extensions to MSBuild is like this. Correctness trumps size.

Reply:
I suppose if the behavior was no-op on wrong msbuild runtime, then you could just reference both if you use both. I find that less pit-of-success but at least that would allow for a good experience when switching between msbuild runtimes.

Reply:
@safern `$(NugetPackageRoot)` will only work if the package is not in any fallback package folder.  If the package is in a fallback folder nuget will not download it to NugetPackageRoot and instead you'll see that it generates a props file with the path to the fallback folder.

Reply:
There is GeneratePathProperty=true metadata on PackageReference which generates $(PkgPackage_Name_With_Dots_As_Underscores ) in *.g.props. That works with fallback folders.

Reply:
https://github.com/NuGet/NuGet.Client/pull/2271

Reply:
That's a pretty cool feature.  I hadn't realized it existed.  I can. think of a handful of places where we could fix some build complexity by using this.  Thanks for the pointer

Reply:
I'll look at updating the logic to use the property suggested by @nguerrera

Reply:
Looks like restore is currently happening with NuGet 4.8.0 and you need 4.9.0 to use `GeneratePathProperty`, looking to see if I can bump the version being used.

Reply:
Should be done, I just needed to update the dotnet SDK to 2.2.103